### PR TITLE
SPOI-11619 Batch support for JDBC input operator

### PR DESCRIPTION
**Pull request by ajaygit158**
@bhupeshchawda Kindly review

### CODE REVIEW

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/JdbcPOJOInputOperator.java [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+ * {@link #queryToRetrieveData()} and {@link #setRuntimeParams()}.
+ *
+ * This class is an exact copy from malhar with the addition of getOutputPort().
+ * Also, getOutputPort() in emitTuples() instead of the variable outputPort
+ *
+ * @displayName Jdbc Input Operator
+ * @category Input
+ * @tags database, sql, pojo, jdbc
+ */
+@org.apache.hadoop.classification.InterfaceStability.Evolving
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/JdbcPOJOInputOperator.java]: Are there any changes in this class? Or is this exactly as is from malhar with just methods made accessible?

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    implements Operator.CheckpointNotificationListener, Batchable
+{
+
+  protected boolean startBatchEmitted = false;
+  protected boolean endBatchEmitted = false;
+  protected transient long recordsEmittedAtEmitTuplesStart;
+  private long shutdownWindowId = -1;
+  private transient long currentWindowId;
+
+  ControlAwareDefaultOutputPort<Object> outputPort = new ControlAwareDefaultOutputPort<Object>()
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: missing schemaAware?

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    currentWindowId = windowId;
+  }
+
+  @Override
+  public void emitTuples()
+  {
+    recordsEmittedAtEmitTuplesStart = tuplesRead;
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: why is this needed?

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  /**
+   * decides and returns whether end batch control tuple should be emitted or
+   * not
+   *
+   * @return
+   */
+  protected boolean shouldEmitEndBatch()
+  {
+    //If no tuples were emitted in the emitTuples call
+    return (!endBatchEmitted && recordsEmittedAtEmitTuplesStart == tuplesRead);
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: Is it possible that this detects the end of a single window worth of data and not that of a batch?

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+      endBatchEmitted = true;
+      shutdownWindowId = this.currentWindowId;
+    }
+  }
+
+  @Override
+  public void emitStartBatchControlTuple()
+  {
+    BatchControlTuple startBatchControlTuple = new BatchControlTuple.StartBatchControlTupleImpl();
+    ControlAwareDefaultOutputPort<Object> output = (ControlAwareDefaultOutputPort<Object>)getOutputPort();
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: can we do an instanceof check here before casting?

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    output.emitControl(startBatchControlTuple);
+  }
+
+  @Override
+  public void emitEndBatchControlTuple()
+  {
+    BatchControlTuple endBatchControlTuple = new BatchControlTuple.EndBatchControlTupleImpl();
+    ControlAwareDefaultOutputPort<Object> output = (ControlAwareDefaultOutputPort<Object>)getOutputPort();
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: same here?

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/JdbcPOJOInputOperator.java (hunk 1 above)
Comment (ajaygit158) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/JdbcPOJOInputOperator.java]: It is the same class with additional getOutputPort() method. I will remove Evolving.

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java (hunk 4 above)
Comment (ajaygit158) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: You are right, I forgot to consider fetchSize. I will change the logic to check tuples emitted in one window instead of one emitTuples call.

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java (hunk 3 above)
Comment (ajaygit158) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: For end batch condition, if no more tuples are available we will end the batch. It simply stores the tuples emitted till the time emitTuples was called. If no tuples are emitted, we can assume end batch. However, as mentioned in another comment below, we will need to change this logic to consider an entire window instead of single emitTuples call.

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java (hunk 2 above)
Comment (ajaygit158) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: Shouldn't we make this as a separate PR?

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+      emitEndBatchControlTuple();
+      endBatchEmitted = true;
+      shutdownWindowId = this.currentWindowId;
+    }
+  }
+
+  @Override
+  public void emitStartBatchControlTuple()
+  {
+    BatchControlTuple startBatchControlTuple = new BatchControlTuple.StartBatchControlTupleImpl();
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: This can also go inside `if`.

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    if (getOutputPort() instanceof ControlAwareDefaultOutputPort) {
+      ControlAwareDefaultOutputPort<Object> output = (ControlAwareDefaultOutputPort<Object>)getOutputPort();
+      output.emitControl(startBatchControlTuple);
+    }
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: else log.error()..?

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  @Override
+  public void emitEndBatchControlTuple()
+  {
+    BatchControlTuple endBatchControlTuple = new BatchControlTuple.EndBatchControlTupleImpl();
```
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/BatchJdbcPOJOInputOperator.java]: same as above..

File: operators/library/src/main/java/com/datatorrent/io/db/jdbc/JdbcPOJOInputOperator.java (hunk 1 above)
Comment (bhupeshchawda) [inline operators/library/src/main/java/com/datatorrent/io/db/jdbc/JdbcPOJOInputOperator.java]: I think better to keep evolving

**Comment by ajaygit158** [comment]
@bhupeshchawda Updated the code with changes suggeted.
